### PR TITLE
Readd Constructors, Modules, Iterators and Generators, Properties, Variables to Readme.md

### PR DESCRIPTION
**Pull request by TimDaub**
Welcome to part 3.

Other parts
- #1 
- #2

If you're wondering what this is all about, read [this](https://github.com/ascribe/javascript#introduction) and [this](https://github.com/ascribe/javascript#approach).

**Feel free to ask questions, leave notes and propose changes.**

[To see a pretty view of the diff (takes you to the first readded Chapter of this PR)](https://github.com/ascribe/javascript/pull/3/files?short_path=04c6e90&unchanged=expanded#9.1)
[To see the file in the branch (more readable)](https://github.com/ascribe/javascript/blob/Review-Constructors-Modules-Iterators_and_Generators-Properties-Variables/README.md#constructors)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -700,4 +700,342 @@ Other Style Guides
...
+    }
+
+    const luke = new Jedi();
+
+    luke.jump()
+      .setHeight(20);
+    ```
+
+
+  - [9.4](#9.4) <a name='9.4'></a> It's okay to write a custom toString() method, just make sure it works successfully and causes no side effects.
```
Comment (sohkai) [inline README.md:794]: Worthwhile trivia: Javascript actually uses the .toString() and .valueOf() when it needs to convert between types. The entire process maybe has too many rules, but it's worth knowing that providing your own .toString() lets you convert the object to a string with the `+` operator like you would if you had operator overloading:

``` javascript
var coolKid = { toString: () => 'Bob'  }
console.log("The cool kid's name is " + coolKid);    // The cool kid's name is Bob
```

Full details: http://speakingjs.com/es5/ch08.html#toprimitive

File: README.md [hunk 2]
```diff
@@ -700,4 +700,342 @@ Other Style Guides
...
+
+    ```javascript
+    class Jedi {
+      constructor(options = {}) {
```
Comment (vrde) [inline README.md]: why not `constructor({ name }) { ... }`? Should we use [destructuring](https://github.com/ascribe/javascript#destructuring) in this case or not?

Comment (TimDaub) [inline README.md]: Superior to this example would be:

``` javascript
class Jedi {
  constructor({ name =  'no name' } = {}, additionalParam) {
    this.name = name;
  }
}
```

I'd say we shouldn't specify how people write classes. Because depending on the classes use case you compose your class signature out of what this style guide provides, which is:
- default parameters
- function signature destructuring (includes default + named parameters)
- vanilla parameters

File: README.md [hunk 3]
```diff
@@ -700,4 +700,342 @@ Other Style Guides
...
+    }
+
+    sum === 15;
+
+    // good
+    let sum = 0;
+    numbers.forEach((num) => sum += num);
+    sum === 15;
+
+    // best (use the functional force)
```
Comment (TimDaub) [inline README.md]: Since some examples are already in Star Wars context, I'd like to add `, Luke` to this comment :laughing: 

`// best (use the functional force, Luke)`

**Comment by TimDaub** [comment]
Read it and actually didn't have too much to complain, except for that one line @vrde pointed out. With that adjusted for me it's fine :+1:

**Comment by rhsimplex** [comment]
> 11.1 Don't use iterators. Prefer JavaScript's higher-order functions like map() and reduce() instead of loops like for-of.

What?? I know this is just a guide, but `map` has no terminating condition and "Don't use iterators" is a pretty strong statement; where performance is concerned, this could mean a lot of wasted iterations.  By all means, use `map` when, well, mapping. But for scanning a list...?

How about:

> 11.1 Prefer JavaScript's higher-order functions like map() and reduce() instead of loops like for-of, unless there is a substantial performance disadvantage by doing so.

slj

**Comment by TimDaub** [comment]
> What?? I know this is just a guide, but map has no terminating condition and "Don't use iterators" is a pretty strong statement; where performance is concerned, this could mean a lot of wasted iterations.

You're absolutely right!
However, to me its also unclear what they're trying to say here.
It could very well be that they're discouraging the usage of "Javascript iterators" by what they mean the

``` javascript
for(let x of list) {}
```

construct while still allowing plain Javascript for-loops:

``` javascript
for(let x = 0; x < list.length; x++) {}
```

> By all means, use map when, well, mapping. But for scanning a list...?

Yes, totally agree. Use `map` for mapping, and so on.... However, for scanning a list there might also
be higher-order functions of an `array` that allow you to avoid using "Javascript iterators" + having nice performance:

-[Array.prototype.every(): tests whether all elements in the array pass the test implemented by the provided function](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/every)

-[Array.prototype.find(): returns a value in the array, if an element in the array satisfies the provided testing function. Otherwise undefined is returned](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)

-[Array.prototype.findIndex(): returns an index in the array, if an element in the array satisfies the provided testing function. Otherwise -1 is returned](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)

-[Array.prototype.includes(): determines whether an array includes a certain element, returning true or false as appropriate](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes)

-[Array.prototype.includes(): determines whether an array includes a certain element, returning true or false as appropriate](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes)

... [(there is a lot more...)](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype)

@all maybe we can agree upon what is meant by "Javascript iterators" and then define what can be done with them.

For me:
1. favor plain Javascript for-loops over Javascript iterators
2. favor higher-order functions over Javascript for-loops

@rhsimplex: Maybe you also want to go to their repo and ask them nicely: https://github.com/airbnb/javascript

**Comment by rhsimplex** [comment]
Ah ok, didn't know about these other functions (shocking, I know). But they should cover virtually every case.  Maybe include a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array then?

**Comment by vrde** [comment]
@TimDaub 

> For me:
> 1. favor plain Javascript for-loops over Javascript iterators
> 2. favor higher-order functions over Javascript for-loops

What's wrong with the `for (let x of list) { ... }`? Why it's worse than `for (let i = 0; i < list.length; i++) { ... }`?

**Comment by TimDaub** [comment]
> What's wrong with the for (let x of list) { ... }? Why it's worse than for (let i = 0; i < list.length; i++) { ... }?

Oh, nothing. That was just me trying to express personal taste. I'd be totally fine with the other way around :D
Also: If we favor `for-of` we need to make it very clear that `for-in` is the devil.

**Comment by sohkai** [comment]
The for-of loop was brought up previously in an [issue on the airbnb styleguide](https://github.com/airbnb/javascript/issues/385).

TLDR; they haven't found much use for it yet, and I suspect they just want to restrict the number of loop constructs that all kind of mean the same thing.

Personally I haven't found much use for it either; typically a .forEach() will suffice unless you need the loop control at which point we can decide whether or not to use the normal for loop or for-of loop. Even then, if you need loop controls, you'll likely want the index anyway so the for-of is useless. Moreover, using for-of requires the dreaded babel polyfill (btw, we still need this polyfill since we're using a Set in one of our utils... but that's the only other instance I know of where we still need the polyfill).

@TimDaub I was pretty convinced for-in has always been the devil :japanese_ogre:.

File: README.md (hunk 1 above)
Comment (cdemirkiran) [inline README.md:794]: Why should we always avoid prototypes? Isn't it a bit too hardcore?
http://thecodeship.com/web-development/methods-within-constructor-vs-prototype-in-javascript/
As far as I know they use less memory which is good because it allows you to save memory.
This is important because JS is used in frontend.

**Comment by cdemirkiran** [comment]
Why are we adding const for each var defined?
By the way where is our let?
const is for constant as far as i know or not?

**Comment by TimDaub** [comment]
Guys, I included all your given feedback.
